src/celeritas/phys/PhysicsStepUtils.hh

```diff
@@ -313,31 +358,14 @@ select_discrete_interaction(const ParticleTrackView& particle,
         [&pstep](ParticleProcessId ppid) { return pstep.per_process_xs(ppid); },
         ParticleProcessId{physics.num_particle_processes()},
         pstep.macro_xs())(rng);
+    CELER_ASSERT(ppid);
 
-    // Determine if the discrete interaction occurs for energy loss
-    // processes
-    if (physics.use_integral_xs(ppid))
+    if (physics.eloss_ppid()
+        && reject_after_eloss(particle.energy(), physics, pstep, ppid, rng))
     {
-        // This is an energy loss process that was sampled for a
-        // discrete interaction, so it will have macro xs tables
-        auto grid_id = physics.value_grid(ValueGridType::macro_xs, ppid);
-        CELER_ASSERT(grid_id);
-
-        // Recalculate the cross section at the post-step energy \f$
-        // E_1 \f$
-        auto      calc_xs = physics.make_calculator<XsCalculator>(grid_id);
-        real_type xs      = calc_xs(particle.energy());
-
-        // The discrete interaction occurs with probability \f$ \sigma(E_1) /
-        // \sigma_{\max} \f$. Note that it's possible for \f$ \sigma(E_1) \f$
-        // to be larger than the estimate of the maximum cross section over the
-        // step \f$ \sigma_{\max} \f$.
-        if (generate_canonical(rng) * pstep.per_process_xs(ppid) > xs)
-        {
-            // No interaction occurs; reset the physics state and continue
-            // tracking
-            return physics.scalars().integral_rejection_action();
-        }
+        // Particle energy has changed over the step and the interaction is no
+        // longer valid
+        return physics.scalars().integral_rejection_action();
```

Comment:
Yes, that's correct. It's not optimal: we should be constructing the model limits based on the process cross sections we're given (or the other way around if we generate cross sections ourselves rather than import them from geant4). But I didn't want to open that can of worms.